### PR TITLE
Validate JavaScript condition while saving (#12901)

### LINKED ISSUE #12901: Layers - Invalid javascript condition
### Discussed in https://github.com/OrchardCMS/OrchardCore/discussions/12894

<div type='discussions-op-text'>

<sup>Originally posted by **sobotama** November 28, 2022</sup>

### Describe the bug
One of our users writes javascript conditions in layer rules for resolving URLs and once they mistyped and wrote the condition like this: 

`("test")`

instead of this:

`url("test")`

After adding any widget to this faulty layer the whole frontend fails with code 500: 

```
System.FormatException: String 'test' was not recognized as a valid Boolean.
   at System.Boolean.Parse(ReadOnlySpan1 value)
   at System.Boolean.Parse(String value)
   at System.String.System.IConvertible.ToBoolean(IFormatProvider provider)
   at System.Convert.ToBoolean(Object value)
   at OrchardCore.Rules.Services.JavascriptConditionEvaluator.EvaluateAsync(JavascriptCondition condition) in src\OrchardCore.Modules\OrchardCore.Rules\Services\JavascriptConditionEvaluator.cs:line 30
   at OrchardCore.Rules.ConditionEvaluator1.OrchardCore.Rules.IConditionEvaluator.EvaluateAsync(Condition condition) in src\OrchardCore\OrchardCore.Rules.Abstractions\ConditionEvaluator.cs:line 16
   at OrchardCore.Rules.Services.RuleService.EvaluateAsync(Rule rule) in src\OrchardCore.Modules\OrchardCore.Rules\Services\RuleService.cs:line 20
   at OrchardCore.Layers.Services.LayerFilter.OnResultExecutionAsync(ResultExecutingContext context, ResultExecutionDelegate next) in src\OrchardCore.Modules\OrchardCore.Layers\Services\LayerFilter.cs:line 113
the rest of stack is ommited.
```

Backend is still accesible and it can be repaired, however the client is unhappy, that the whole site can be broken this easily, even though they now use the proper URL condition. 

### To Reproduce
1. Create a layer
2. Add rule - Javascript condition with this script: `("test")` - or any other mistake - `url("test)`, `ul("test")`
3. Assign any widget to the layer
4. Go to anywhere in the frontend

### Expected behavior
Maybe just not showing the affected layer and logging the error, while the rest of the site still functions normally

I tried it on OrchardCore versions 1.1, 1.4 and 1.5 with all of them behaving the same.
</div>

### PR DESCRIPTION
**Pull request by sobotama**
Fix #12901

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs
Comment (ns8482e) [inline src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs:74]: use ParseScript instead of evaluate

Comment (sobotama) [reply src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs:74]: Should I parse it first and then evaluate? Sébastien Roses second point in the issue is "ensure that evaluating the script is not breaking the full page..." so I have to evaluate and I thought that since EvaluateAsync parses it with ParseScript before evaluation then I only need to catch the exception, otherwise I would be parsing it twice.

**Comment by Piedone** [comment]
I sincerely apologize for us taking so much time here. I checked out the PR, going over old ones, and this would be quite useful. It needs some tweaks, and the merge conflict resolved though. And foremost, so we can do anything with the code, you'd need to accept the CLA (unfortunately, the CLA bot was broken for a while).

Is this something you'd like to revisit any time soon @sobotama or should we close?

**Comment by sobotama** [comment]
Hello,

thanks for responding, I completely forgot that this was still unresovled.

I took care of the merge conflict, I'm currently figuring out the CLA with the company I work at and in the meantime, is there something else I should do @Piedone?

**Comment by sobotama** [comment]
@dotnet-policy-service agree

**Comment by Piedone** [comment]
Great, thank you! @ns8482e would you like to get back to the review, or should I take over?

**Comment by Piedone** [comment]
@ns8482e could you reply here, please?

File: src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs:62]:
```suggestion
                updater.ModelState.AddModelError(Prefix, nameof(model.Script), S["Please provide a script."]);
```

File: src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs:63]:
```suggestion
```

File: src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs:60]: Let's move this general comment above the whole code block where it makes sense.

```suggestion
            // CodeMirror hides the textarea which displays the error when updater.ModelState.AddModelError() is used,
            // that's why a notifier is used to show validation errors.
            if (string.IsNullOrWhiteSpace(model.Script))
```

File: src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Rules/Drivers/JavascriptConditionDisplayDriver.cs:64]:
```suggestion
                await _notifier.ErrorAsync(H["Please provide a script."]);
```

**Review by Piedone — CHANGES_REQUESTED** [review]
Let me take over the review here, and thank you for your patience! I have some small remarks; once you address those, I'll merge ASAP.

When addressing review feedback, please adhere to the following:

- Please update your pull request according to feedback until it is approved by one of the core team members.
- Apply [suggested changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) directly so the reviewer doesn't have to eyeball the changes. These resolve themselves after applying them, and that's fine.
- Don't resolve other conversations so it's easier to track for the reviewer. Then, the reviewer will resolve them.
- Feel free to mark conversations that you addressed to keep track of them with an emoji or otherwise, just don't resolve them.
- Please keep conversations happening in line comments in those convos, otherwise, communication will be a mess. If you have trouble finding them, see [this video](https://github.com/OrchardCMS/OrchardCore/pull/14749#issuecomment-1917976028).
- When you're done addressing all feedback of a review, click ["Re-request review"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#re-requesting-a-review) in the top-right corner for each reviewer when you're ready for another round of review, so they know that you're done.

**Review by Piedone — APPROVED** [review]

**Comment by Piedone** [comment]
@allcontributors please add @sobotama for code

**Comment by Piedone** [comment]
Thank you, and congrats on your first contribution to Orchard Core! I promise the next one won't take us two years to accept :). We implemented a lot of improvements in managing PRs since then: https://github.com/OrchardCMS/OrchardCore/issues/15029.

Once the docs page build finishes, you'll also show up in the hall of fame here: https://docs.orchardcore.net/en/latest/docs/community/contributors/.

**Comment by Piedone** [comment]
Ah, I forgot that adding contributors is now automated :).

**Pull request merged**
